google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/ConnectionImpl.java

```diff
@@ -623,6 +633,55 @@ public void setStatementTag(String tag) {
     this.statementTag = tag;
   }
 
+  @Override
+  public byte[] getProtoDescriptors() {
+    ConnectionPreconditions.checkState(!isClosed(), CLOSED_ERROR_MSG);
+    if (this.protoDescriptorsFilePath != null) {
```

Comment:
I think it's better to do this slightly differently:
1. The check should be `this.protoDescriptors == null && this.protoDescriptorsFilePath != null`
2. At the end of this method, do not clear the field `this.protoDescriptorsFilePath`
3. In the setter methods for Proto Descriptors, we should make sure to clear both fields when one of them is set.

The advantage of the above is that you retain the fact that the current proto descriptors were set using a file name. That gives you more opportunities to include the file name in any error message that we might need to return in the future (for example if it turns out that the proto descriptor was invalid).

Reply:
Thanks for the suggestion, Knut. This makes more sense. I have modified the logic as above.